Error linking Scala.js due to java.util.zip.ZipException: zip END header not found

**Version(s)**

1.0.0-RC2 and 1.0.0

**Describe the bug**

I just started to experiment with scala-js and scala-cli and noticed that the package command could fail with an error in linking Scala.js becuase a ` java.util.zip.ZipException: zip END header not found` was thrown.

**To Reproduce**

Given the following file, named App.scala:
```scala
//> using scala 3.3.0

//> using platform scala-js
//> using jsVersion 1.13.1

//> using options -Wvalue-discard -Wunused:all

object App:
  @main def run(): Unit =
    println("hello world")
```

And I package my application with: the `scala-cli --power package . --output App.js --force` command. The first attempt on a clean project always succeeds, but if change something in the code, e.g.: the `println("hello world")` to `println("hello")` the next package command will fail with error. Here is the verbose output of such failure:

```
Checking for a running Bloop server at /Users/suliatis/Library/Caches/ScalaCli/bloop/daemon ...
Attempting to connect to Bloop server /Users/suliatis/Library/Caches/ScalaCli/bloop/daemon ...
Connection attempt result: Some(Right(java.nio.channels.SocketChannel[connected local= remote=/Users/suliatis/Library/Caches/ScalaCli/bloop/daemon/socket]))
nailgun debug: Sending arguments '' to Nailgun server
nailgun debug: Sending environment variables to Nailgun server
nailgun debug: Sending working directory /Users/suliatis/code/test to Nailgun server
nailgun debug: Sending command to about Nailgun server
nailgun debug: Finished sending command information to Nailgun server
nailgun debug: Starting thread to read stdin...
nailgun debug: Received action Print([B@17602792) from Nailgun server
nailgun debug: Received action Print([B@59ef9600) from Nailgun server
nailgun debug: Received action Print([B@b570169) from Nailgun server
nailgun debug: Received action Print([B@356a195c) from Nailgun server
nailgun debug: Received action Print([B@3f253c85) from Nailgun server
nailgun debug: Received action Print([B@4ea3bd12) from Nailgun server
nailgun debug: Received action Exit(0) from Nailgun server
nailgun debug: Waiting for stdin thread to finish...
nailgun debug: Waiting for heartbeat thread to finish...
nailgun debug: Returning exit code...
Checking for a running Bloop server at /Users/suliatis/Library/Caches/ScalaCli/bloop/daemon ...
Attempting to connect to Bloop server /Users/suliatis/Library/Caches/ScalaCli/bloop/daemon ...
Connection attempt result: Some(Right(java.nio.channels.SocketChannel[connected local= remote=/Users/suliatis/Library/Caches/ScalaCli/bloop/daemon/socket]))
Checking for a running Bloop server at /Users/suliatis/Library/Caches/ScalaCli/bloop/daemon ...
Attempting to connect to Bloop server /Users/suliatis/Library/Caches/ScalaCli/bloop/daemon ...
Connection attempt result: Some(Right(java.nio.channels.SocketChannel[connected local= remote=/Users/suliatis/Library/Caches/ScalaCli/bloop/daemon/socket]))
nailgun debug: Sending arguments '' to Nailgun server
nailgun debug: Sending environment variables to Nailgun server
nailgun debug: Sending working directory /Users/suliatis/code/test to Nailgun server
nailgun debug: Sending command to about Nailgun server
nailgun debug: Finished sending command information to Nailgun server
nailgun debug: Starting thread to read stdin...
nailgun debug: Received action Print([B@2b0e5787) from Nailgun server
nailgun debug: Received action Print([B@3c3287aa) from Nailgun server
nailgun debug: Received action Print([B@2db8f4f0) from Nailgun server
nailgun debug: Received action Print([B@17e18f8) from Nailgun server
nailgun debug: Received action Print([B@1c0c70c1) from Nailgun server
nailgun debug: Received action Print([B@a9b99f7) from Nailgun server
nailgun debug: Received action Exit(0) from Nailgun server
nailgun debug: Waiting for stdin thread to finish...
nailgun debug: Waiting for heartbeat thread to finish...
nailgun debug: Returning exit code...
Opening BSP connection with bloop
Bloop BSP connection waiting at local:/Users/suliatis/Library/Caches/ScalaCli/bsp-sockets/proc-32519
BSP connection at /Users/suliatis/Library/Caches/ScalaCli/bsp-sockets/proc-32519 not found, waiting 100 milliseconds
nailgun debug: Sending arguments '--protocol local --socket /Users/suliatis/Library/Caches/ScalaCli/bsp-sockets/proc-32519' to Nailgun server
nailgun debug: Sending environment variables to Nailgun server
nailgun debug: Sending working directory /Users/suliatis/code/test/.scala-build to Nailgun server
nailgun debug: Sending command to bsp Nailgun server
nailgun debug: Finished sending command information to Nailgun server
nailgun debug: Starting thread to read stdin...
nailgun debug: Received action Print([B@705efb9) from Nailgun server
The server is listening for incoming connections at local:///Users/suliatis/Library/Caches/ScalaCli/bsp-sockets/proc-32519...
BSP connection at /Users/suliatis/Library/Caches/ScalaCli/bsp-sockets/proc-32519 opened
Connected to Bloop via BSP at local:/Users/suliatis/Library/Caches/ScalaCli/bsp-sockets/proc-32519
Connected to Bloop via BSP at local:/Users/suliatis/Library/Caches/ScalaCli/bsp-sockets/proc-32519
nailgun debug: Received action Print([B@518ac20d) from Nailgun server
Accepted incoming BSP client connection at local:///Users/suliatis/Library/Caches/ScalaCli/bsp-sockets/proc-32519
Sending buildInitialize BSP command to Bloop
nailgun debug: Received action Print([B@232c39c7) from Nailgun server
request received: build/initialize
nailgun debug: Received action Print([B@1aa67542) from Nailgun server
BSP initialization handshake complete.
Fetching List(Dependency(org.scala-lang:scala3-compiler_3, 3.3.0, Configuration(), Set(), Publication(, Type(), Extension(), Classifier()), false, true)), adding List(Ivy
Reposi
tory(Pattern(List(Const(file:///Users/suliatis/Library/Caches/ScalaCli/local-repo/v1.0.0//), Var(organisation), Const(/), Var(module), Const(/), Opt(List(Const(scala_), V
ar(sca
laVersion), Const(/))), Opt(List(Const(sbt_), Var(sbtVersion), Const(/))), Var(revision), Const(/), Var(type), Const(s/), Var(artifact), Opt(List(Const(-), Var(classifier
))), C
onst(.), Var(ext))), None, None, true, true, true, false, None, true))
Found 13 artifacts:
  /Users/suliatis/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala3-compiler_3/3.3.0/scala3-compiler_3-3.3.0.jar
  /Users/suliatis/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala3-interfaces/3.3.0/scala3-interfaces-3.3.0.jar
  /Users/suliatis/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala3-library_3/3.3.0/scala3-library_3-3.3.0.jar
  /Users/suliatis/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/tasty-core_3/3.3.0/tasty-core_3-3.3.0.jar
  /Users/suliatis/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/modules/scala-asm/9.4.0-scala-1/scala-asm-9.4.0-scala-1.jar
  /Users/suliatis/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-sbt/compiler-interface/1.3.5/compiler-interface-1.3.5.jar
  /Users/suliatis/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/jline/jline-reader/3.19.0/jline-reader-3.19.0.jar
  /Users/suliatis/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/jline/jline-terminal/3.19.0/jline-terminal-3.19.0.jar
  /Users/suliatis/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/jline/jline-terminal-jna/3.19.0/jline-terminal-jna-3.19.0.jar
  /Users/suliatis/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-library/2.13.10/scala-library-2.13.10.jar
  /Users/suliatis/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/com/google/protobuf/protobuf-java/3.7.0/protobuf-java-3.7.0.jar
  /Users/suliatis/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-sbt/util-interface/1.3.0/util-interface-1.3.0.jar
  /Users/suliatis/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/net/java/dev/jna/jna/5.3.1/jna-5.3.1.jar

Fetching List(Dependency(org.virtuslab.scala-cli:scalajscli_2.13, 1.13.1+, Configuration(), Set(), Publication(, Type(), Extension(), Classifier()), false, true)), adding
 List(
IvyRepository(Pattern(List(Const(file:///Users/suliatis/Library/Caches/ScalaCli/local-repo/v1.0.0//), Var(organisation), Const(/), Var(module), Const(/), Opt(List(Const(s
cala_)
, Var(scalaVersion), Const(/))), Opt(List(Const(sbt_), Var(sbtVersion), Const(/))), Var(revision), Const(/), Var(type), Const(s/), Var(artifact), Opt(List(Const(-), Var(c
lassif
ier))), Const(.), Var(ext))), None, None, true, true, true, false, None, true))
Fetching List(Dependency(org.scala-js:scalajs-library_2.13, 1.13.1, Configuration(), Set(), Publication(, Type(), Extension(), Classifier()), false, true), Dependency(org
.scala
-lang:scala3-library_sjs1_3, 3.3.0, Configuration(), Set(), Publication(, Type(), Extension(), Classifier()), false, true)), adding List(IvyRepository(Pattern(List(Const(
file:/
//Users/suliatis/Library/Caches/ScalaCli/local-repo/v1.0.0//), Var(organisation), Const(/), Var(module), Const(/), Opt(List(Const(scala_), Var(scalaVersion), Const(/))), 
Opt(Li
st(Const(sbt_), Var(sbtVersion), Const(/))), Var(revision), Const(/), Var(type), Const(s/), Var(artifact), Opt(List(Const(-), Var(classifier))), Const(.), Var(ext))), Non
e, Non
e, true, true, true, false, None, true))
Checking Bloop project in /Users/suliatis/code/test/.scala-build/.bloop/project_04b1c71d63.json
Bloop project in /Users/suliatis/code/test/.scala-build/.bloop/project_04b1c71d63.json doesn't need updating
Listing BSP build targets
Compiling project_04b1c71d63 with Bloop
Received onBuildTaskStart from bloop: TaskStartParams [
  taskId = TaskId [
    id = "1"
    parents = null
  ]
  eventTime = 1685291783091
  message = "Start no-op compilation for project_04b1c71d63"
  dataKind = "compile-task"
  data = {"target":{"uri":"file:/Users/suliatis/code/test/.scala-build/?id=project_04b1c71d63"}}
]
Received onBuildTaskFinish from bloop: TaskFinishParams [
  taskId = TaskId [
    id = "1"
    parents = null
  ]
  eventTime = 1685291783095
  message = "Compiled 'project_04b1c71d63'"
  status = OK
  dataKind = "compile-report"
  data = {"target":{"uri":"file:/Users/suliatis/code/test/.scala-build/?id=project_04b1c71d63"},"errors":0,"warnings":0,"isNoOp":true,"isLastCycle":true,"clientDir":"file
:///Us
ers/suliatis/code/test/.scala-build/project_04b1c71d63/classes/main/","analysisOut":"file:///Users/suliatis/code/test/.scala-build/.bloop/project_04b1c71d63/project_04b1c
71d63-
analysis.bin"}
]
Compilation succeeded
Post-processing class files of pre-processed sources
Moving semantic DBs around
nailgun debug: Received action Print([B@5ac72fc4) from Nailgun server
BSP server stopped
nailgun debug: Received action Exit(0) from Nailgun server
nailgun debug: Waiting for stdin thread to finish...
nailgun debug: Waiting for heartbeat thread to finish...
nailgun debug: Returning exit code...
Getting https://github.com/virtusLab/scala-js-cli/releases/download/v1.13.1/scala-js-ld-aarch64-apple-darwin.gz
https://github.com/virtusLab/scala-js-cli/releases/download/v1.13.1/scala-js-ld-aarch64-apple-darwin.gz not found (coursier.cache.ArtifactError$NotFound: not found: https
://git
hub.com/virtusLab/scala-js-cli/releases/download/v1.13.1/scala-js-ld-aarch64-apple-darwin.gz)
Could not fetch binary, fetching JVM dependencies:
  org.virtuslab.scala-cli:scalajscli_2.13:1.13.1+

  Running
/Users/suliatis/Library/Caches/Coursier/arc/https/github.com/adoptium/temurin17-binaries/releases/download/jdk-17.0.5%252B8/OpenJDK17U-jdk_aarch64_mac_hotspot_17.0.5_8.ta
r.gz/j
dk-17.0.5+8/Contents/Home/bin/java
-cp
/Users/suliatis/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/virtuslab/scala-cli/scalajscli_2.13/1.13.1/scalajscli_2.13-1.13.1.jar:/Users/suliatis/Library/
Caches
/Coursier/v1/https/repo1.maven.org/maven2/org/scala-js/scalajs-linker_2.13/1.13.1/scalajs-linker_2.13-1.13.1.jar:/Users/suliatis/Library/Caches/Coursier/v1/https/repo1.ma
ven.or
g/maven2/com/github/scopt/scopt_2.13/4.1.0/scopt_2.13-4.1.0.jar:/Users/suliatis/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-library/2.13.
10/sca
la-library-2.13.10.jar:/Users/suliatis/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-js/scalajs-linker-interface_2.13/1.13.1/scalajs-linker-interface_
2.13-1
.13.1.jar:/Users/suliatis/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-js/scalajs-ir_2.13/1.13.1/scalajs-ir_2.13-1.13.1.jar:/Users/suliatis/Library/C
aches/
Coursier/v1/https/repo1.maven.org/maven2/com/google/javascript/closure-compiler/v20220202/closure-compiler-v20220202.jar:/Users/suliatis/Library/Caches/Coursier/v1/https/
repo1.
maven.org/maven2/org/scala-lang/modules/scala-parallel-collections_2.13/0.2.0/scala-parallel-collections_2.13-0.2.0.jar:/Users/suliatis/Library/Caches/Coursier/v1/https/r
epo1.m
aven.org/maven2/org/scala-js/scalajs-logging_2.13/1.1.1/scalajs-logging_2.13-1.1.1.jar
org.scalajs.cli.Scalajsld
--outputDir
/var/folders/3g/2djfyqxs1950k93cwhjyzs_00000gn/T/scala-cli-js-linking9447955392956969430
--mainMethod
run.main
--fastOpt
--compliantAsInstanceOfs
--moduleKind
NoModule
--moduleSplitStyle
FewestModules
--es2015
/Users/suliatis/code/test/.scala-build/project_04b1c71d63/jar/library.jar
/Users/suliatis/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-js/scalajs-library_2.13/1.13.1/scalajs-library_2.13-1.13.1.jar
/Users/suliatis/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala3-library_sjs1_3/3.3.0/scala3-library_sjs1_3-3.3.0.jar
/Users/suliatis/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-lang/scala-library/2.13.10/scala-library-2.13.10.jar
/Users/suliatis/Library/Caches/Coursier/v1/https/repo1.maven.org/maven2/org/scala-js/scalajs-javalib/1.13.1/scalajs-javalib-1.13.1.jar

Exception in thread "main" java.util.zip.ZipException: invalid CEN header (bad signature)
        at jdk.zipfs/jdk.nio.zipfs.ZipFileSystem.initCEN(ZipFileSystem.java:1562)
        at jdk.zipfs/jdk.nio.zipfs.ZipFileSystem.<init>(ZipFileSystem.java:179)
        at jdk.zipfs/jdk.nio.zipfs.ZipFileSystemProvider.getZipFileSystem(ZipFileSystemProvider.java:125)
        at jdk.zipfs/jdk.nio.zipfs.ZipFileSystemProvider.newFileSystem(ZipFileSystemProvider.java:120)
        at java.base/java.nio.file.FileSystems.newFileSystem(FileSystems.java:528)
        at java.base/java.nio.file.FileSystems.newFileSystem(FileSystems.java:400)
        at org.scalajs.linker.PathIRContainer$JarIRContainer.$anonfun$sjsirFiles$2(PathIRContainer.scala:65)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
        at scala.concurrent.impl.ExecutionContextImpl$DefaultThreadFactory$$anon$1$$anon$2.block(ExecutionContextImpl.scala:59)
        at java.base/java.util.concurrent.ForkJoinPool.compensatedBlock(ForkJoinPool.java:3449)
        at java.base/java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3432)
        at scala.concurrent.impl.ExecutionContextImpl$DefaultThreadFactory$$anon$1.blockOn(ExecutionContextImpl.scala:70)
        at scala.concurrent.package$.blocking(package.scala:124)
        at org.scalajs.linker.PathIRContainer$JarIRContainer.$anonfun$sjsirFiles$1(PathIRContainer.scala:61)
        at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:678)
        at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:467)
        at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1395)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:162
2)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
Scala.js linker exited with return code 1
Exception in thread "main" java.lang.Exception: scala.build.errors.ScalaJsLinkingError: Error linking Scala.js
        at scala.cli.internal.CliLogger.exit(CliLogger.scala:141)
        at scala.cli.commands.util.CommandHelpers$EitherBuildExceptionOps.orExit(CommandHelpers.scala:18)
        at scala.cli.commands.package0.Package$.runCommand(Package.scala:156)
        at scala.cli.commands.package0.Package$.runCommand(Package.scala:70)
        at scala.cli.commands.ScalaCommand.run(ScalaCommand.scala:364)
        at scala.cli.commands.ScalaCommand.run(ScalaCommand.scala:348)
        at caseapp.core.app.CaseApp.main(CaseApp.scala:150)
        at scala.cli.commands.ScalaCommand.main(ScalaCommand.scala:333)
        at caseapp.core.app.CommandsEntryPoint.main(CommandsEntryPoint.scala:120)
        at scala.cli.ScalaCliCommands.main(ScalaCliCommands.scala:124)
        at scala.cli.ScalaCli$.main0(ScalaCli.scala:203)
        at scala.cli.ScalaCli$.main(ScalaCli.scala:107)
        at scala.cli.ScalaCli.main(ScalaCli.scala)
Caused by: scala.build.errors.ScalaJsLinkingError: Error linking Scala.js
        at scala.cli.internal.ScalaJsLinker$.link$$anonfun$1(ScalaJsLinker.scala:152)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.build.EitherCps$Helper.apply(EitherCps.scala:19)
        at scala.cli.internal.ScalaJsLinker$.link(ScalaJsLinker.scala:154)
        at scala.cli.commands.package0.Package$.linkJs$$anonfun$2(Package.scala:939)
        at scala.build.EitherCps$Helper.apply(EitherCps.scala:19)
        at scala.cli.commands.package0.Package$.linkJs(Package.scala:986)
        at scala.cli.commands.package0.Package$.buildJs(Package.scala:686)
        at scala.cli.commands.package0.Package$.doPackage$$anonfun$1(Package.scala:393)
        at scala.build.EitherCps$Helper.apply(EitherCps.scala:19)
        at scala.cli.commands.package0.Package$.doPackage(Package.scala:521)
        at scala.cli.commands.package0.Package$.runCommand(Package.scala:154)
        ... 10 more
```

If I change the code and run the package command again it might recover, but without any change it fails always. I found a workaround, if delete the .scala-build folder I can recover the package command for the next few attempts.

**Expected behaviour**

The project can be repackaged after any change in the code.

--- a/modules/build/src/test/scala/scala/build/tests/TestInputs.scala
+++ b/modules/build/src/test/scala/scala/build/tests/TestInputs.scala
@@ -14,44 +14,52 @@ import scala.util.Try
 
 final case class TestInputs(
   files: Seq[(os.RelPath, String)],
-  inputArgs: Seq[String]
+  inputArgs: Seq[String] = Seq.empty,
+  forceCwd: Option[os.Path] = None
 ) {
   def withInputs[T](f: (os.Path, Inputs) => T): T =
     withCustomInputs(false, None)(f)
 
-  def fromRoot[T](f: os.Path => T): T =
-    TestInputs.withTmpDir("scala-cli-tests-") { tmpDir =>
-      for ((relPath, content) <- files) {
-        val path = tmpDir / relPath
-        os.write(path, content.getBytes(StandardCharsets.UTF_8), createFolders = true)
-      }
+  def fromRoot[T](f: os.Path => T, skipCreatingSources: Boolean = false): T =
+    TestInputs.withTmpDir("scala-cli-tests-", forceCwd) { tmpDir =>
+      if skipCreatingSources then f(tmpDir)
+      else {
+        for ((relPath, content) <- files) {
+          val path = tmpDir / relPath
+          os.write(path, content.getBytes(StandardCharsets.UTF_8), createFolders = true)
+        }
 
-      f(tmpDir)
+        f(tmpDir)
+      }
     }
 
   def withCustomInputs[T](
     viaDirectory: Boolean,
-    forcedWorkspaceOpt: Option[os.FilePath]
+    forcedWorkspaceOpt: Option[os.FilePath],
+    skipCreatingSources: Boolean = false
   )(
     f: (os.Path, Inputs) => T
   ): T =
-    fromRoot { tmpDir =>
-      val inputArgs0 =
-        if (viaDirectory) Seq(tmpDir.toString)
-        else if (inputArgs.isEmpty) files.map(_._1.toString)
-        else inputArgs
-      val res = Inputs(
-        inputArgs0,
-        tmpDir,
-        forcedWorkspace = forcedWorkspaceOpt.map(_.resolveFrom(tmpDir)),
-        allowRestrictedFeatures = true,
-        extraClasspathWasPassed = false
-      )(using ScalaCliInvokeData.dummy)
-      res match {
-        case Left(err)     => throw new Exception(err)
-        case Right(inputs) => f(tmpDir, inputs)
-      }
-    }
+    fromRoot(
+      { tmpDir =>
+        val inputArgs0 =
+          if (viaDirectory) Seq(tmpDir.toString)
+          else if (inputArgs.isEmpty) files.map(_._1.toString)
+          else inputArgs
+        val res = Inputs(
+          inputArgs0,
+          tmpDir,
+          forcedWorkspace = forcedWorkspaceOpt.map(_.resolveFrom(tmpDir)),
+          allowRestrictedFeatures = true,
+          extraClasspathWasPassed = false
+        )(using ScalaCliInvokeData.dummy)
+        res match {
+          case Left(err)     => throw new Exception(err)
+          case Right(inputs) => f(tmpDir, inputs)
+        }
+      },
+      skipCreatingSources
+    )
 
   def withLoadedBuild[T](
     options: BuildOptions,
@@ -72,9 +80,10 @@ final case class TestInputs(
     bloopConfigOpt: Option[BloopRifleConfig],
     fromDirectory: Boolean = false,
     buildTests: Boolean = true,
-    actionableDiagnostics: Boolean = false
+    actionableDiagnostics: Boolean = false,
+    skipCreatingSources: Boolean = false
   )(f: (os.Path, Inputs, Either[BuildException, Builds]) => T): T =
-    withCustomInputs(fromDirectory, None) { (root, inputs) =>
+    withCustomInputs(fromDirectory, None, skipCreatingSources) { (root, inputs) =>
       val compilerMaker = bloopConfigOpt match {
         case Some(bloopConfig) =>
           new BloopCompilerMaker(bloopConfig, buildThreads.bloop, strictBloopJsonCheck = true)
@@ -103,15 +112,17 @@ final case class TestInputs(
     fromDirectory: Boolean = false,
     buildTests: Boolean = true,
     actionableDiagnostics: Boolean = false,
-    scope: Scope = Scope.Main
+    scope: Scope = Scope.Main,
+    skipCreatingSources: Boolean = false
   )(f: (os.Path, Inputs, Either[BuildException, Build]) => T): T =
     withBuilds(
       options,
       buildThreads,
       bloopConfigOpt,
       fromDirectory,
-      buildTests,
-      actionableDiagnostics
+      buildTests = buildTests,
+      actionableDiagnostics = actionableDiagnostics,
+      skipCreatingSources = skipCreatingSources
     ) {
       (p, i, builds) =>
         f(
@@ -127,11 +138,14 @@ object TestInputs {
   def apply(files: (os.RelPath, String)*): TestInputs =
     TestInputs(files, Nil)
 
-  def withTmpDir[T](prefix: String)(f: os.Path => T): T = {
-    val tmpDir = os.temp.dir(prefix = prefix)
-    try f(tmpDir)
-    finally tryRemoveAll(tmpDir)
-  }
+  def withTmpDir[T](prefix: String, forceCwd: Option[os.Path] = None)(f: os.Path => T): T =
+    forceCwd match {
+      case Some(path) => f(path)
+      case None =>
+        val tmpDir = os.temp.dir(prefix = prefix)
+        try f(tmpDir)
+        finally tryRemoveAll(tmpDir)
+    }
 
   def tryRemoveAll(f: os.Path): Unit =
     try os.remove.all(f)

--- a/modules/cli/src/main/scala/scala/cli/packaging/Library.scala
+++ b/modules/cli/src/main/scala/scala/cli/packaging/Library.scala
@@ -1,6 +1,7 @@
 package scala.cli.packaging
 
 import java.io.OutputStream
+import java.nio.file.StandardOpenOption.{CREATE, TRUNCATE_EXISTING}
 import java.nio.file.attribute.FileTime
 import java.util.jar.{Attributes => JarAttributes, JarOutputStream}
 import java.util.zip.{ZipEntry, ZipOutputStream}
@@ -28,7 +29,11 @@ object Library {
     if (cacheData.changed) {
       var outputStream: OutputStream = null
       try {
-        outputStream = os.write.outputStream(dest, createFolders = true)
+        outputStream = os.write.outputStream(
+          dest,
+          createFolders = true,
+          openOptions = Seq(CREATE, TRUNCATE_EXISTING)
+        )
         writeLibraryJarTo(
           outputStream,
           build,

--- a/modules/cli/src/test/scala/cli/tests/PackageTests.scala
+++ b/modules/cli/src/test/scala/cli/tests/PackageTests.scala
@@ -1,0 +1,78 @@
+package cli.tests
+
+import com.eed3si9n.expecty.Expecty.expect
+
+import java.nio.file.FileSystems
+
+import scala.build.Ops.*
+import scala.build.options.{BuildOptions, InternalOptions}
+import scala.build.tests.TestInputs
+import scala.build.tests.util.BloopServer
+import scala.build.{BuildThreads, Directories, LocalRepo}
+import scala.cli.internal.CachedBinary
+import scala.cli.packaging.Library
+import scala.util.{Properties, Random}
+
+class PackageTests extends munit.FunSuite {
+
+  val buildThreads = BuildThreads.create()
+  def bloopConfig  = BloopServer.bloopConfig
+
+  val extraRepoTmpDir = os.temp.dir(prefix = "scala-cli-tests-extra-repo-")
+  val directories     = Directories.under(extraRepoTmpDir)
+
+  val defaultOptions = BuildOptions(
+    internal = InternalOptions(
+      localRepository = LocalRepo.localRepo(directories.localRepoDir)
+    )
+  )
+
+  /** Fixes - https://github.com/VirtusLab/scala-cli/issues/2166 */
+  test(s"should generate a correct jar library when the project was changed") {
+    TestInputs().fromRoot { root =>
+      val inputs = TestInputs(
+        files = Seq(os.rel / "Hello.scala" ->
+          """//> using platform scala-js
+            |//> using options -Wvalue-discard -Wunused:all
+            |
+            |object Hello extends App {
+            |  println("Hello, World World World")
+            |}""".stripMargin),
+        forceCwd = Some(root)
+      )
+      inputs.withBuild(defaultOptions, buildThreads, Some(bloopConfig)) {
+        (_, _, maybeFirstBuild) =>
+          val firstBuild      = maybeFirstBuild.orThrow.successfulOpt.get
+          val firstLibraryJar = Library.libraryJar(firstBuild)
+          expect(os.exists(firstLibraryJar)) // should create library jar
+
+          // change Hello.scala and recompile
+          os.write.over(
+            root / "Hello.scala",
+            """//> using platform scala-js
+              |//> using options -Wvalue-discard -Wunused:all
+              |
+              |object Hello extends App {
+              |  println("hello")
+              |}""".stripMargin
+          )
+
+          inputs.withBuild(
+            defaultOptions,
+            buildThreads,
+            Some(bloopConfig),
+            skipCreatingSources = true
+          ) {
+            (_, _, maybeSecondBuild) =>
+              val secondBuild = maybeSecondBuild.orThrow.successfulOpt.get
+              val libraryJar  = Library.libraryJar(secondBuild)
+              val fs = // should not throw "invalid CEN header (bad signature)" ZipException
+                FileSystems.newFileSystem(libraryJar.toNIO, null: ClassLoader)
+              expect(fs.isOpen)
+              fs.close()
+          }
+      }
+    }
+  }
+
+}

